modin/core/dataframe/pandas/metadata/dtypes.py

```diff
@@ -12,13 +12,524 @@
+    know_all_names : bool, default: True
+        Whether `known_dtypes` and `cols_with_unknown_dtypes` contain all column names for this dataframe besides those,
+        that are being described by `remaining_dtype`.
+        One can't pass `know_all_names=False` together with `remaining_dtype` as this creates ambiguity
+        on how to interpret missing columns (whether they belong to `remaining_dtype` or not).
+    _schema_is_known : bool, optional
+        Whether `known_dtypes` describe all columns in the dataframe. This parameter intended mostly
+        for internal use.
+    """
+
+    def __init__(
+        self,
+        known_dtypes: Optional[dict[IndexLabel, np.dtype]] = None,
+        cols_with_unknown_dtypes: Optional[list[IndexLabel]] = None,
+        remaining_dtype: Optional[np.dtype] = None,
+        parent_df: Optional["PandasDataframe"] = None,
+        columns_order: Optional[dict[int, IndexLabel]] = None,
+        know_all_names: bool = True,
+        _schema_is_known: Optional[bool] = None,
+    ):
+        if not know_all_names and remaining_dtype is not None:
+            raise RuntimeError(
+                "It's not allowed to pass 'remaining_dtype' and 'know_all_names=False' at the same time."
+            )
+        # columns with known dtypes
+        self._known_dtypes: dict[IndexLabel, np.dtype] = (
+            {} if known_dtypes is None else dict(known_dtypes)
+        )
+        if known_dtypes is not None and len(self._known_dtypes) != len(known_dtypes):
+            raise NotImplementedError(
+                "Duplicated column names are not yet supported by DtypesDescriptor"
+            )
+        # columns with unknown dtypes (they're not described by 'remaining_dtype')
+        if cols_with_unknown_dtypes is not None and len(
+            set(cols_with_unknown_dtypes)
+        ) != len(cols_with_unknown_dtypes):
+            raise NotImplementedError(
+                "Duplicated column names are not yet supported by DtypesDescriptor"
+            )
+        self._cols_with_unknown_dtypes: list[IndexLabel] = (
+            [] if cols_with_unknown_dtypes is None else cols_with_unknown_dtypes
+        )
+        # whether 'known_dtypes' describe all columns in the dataframe
+        if _schema_is_known is None:
+            self._schema_is_known: bool = (
+                len(cols_with_unknown_dtypes) == 0
+                if (
+                    # if 'cols_with_unknown_dtypes' was explicitly specified as an empty list and
+                    # we don't have any 'remaining_dtype', then we assume that 'known_dtypes' are complete
+                    cols_with_unknown_dtypes is not None
+                    and know_all_names
+                    and remaining_dtype is None
+                    and len(self._known_dtypes) > 0
+                )
+                else False
+            )
+        else:
+            self._schema_is_known: bool = _schema_is_known
```

Comment:
It seems that this makes the code easier to read:

```suggestion
        self._schema_is_known: bool = _schema_is_known
        if _schema_is_known is None:
            self._schema_is_known = False
            # if 'cols_with_unknown_dtypes' was explicitly specified as an empty list and
            # we don't have any 'remaining_dtype', then we assume that 'known_dtypes' are complete
            if (
                cols_with_unknown_dtypes is not None
                and know_all_names
                and remaining_dtype is None
                and len(self._known_dtypes) > 0
            ):
                self._schema_is_known = len(cols_with_unknown_dtypes) == 0
```